src/Orleans/Runtime/GrainReferenceRuntime.cs

```diff
@@ -0,0 +1,230 @@
+                        if (rejection == null)
+                        {
+                            if (string.IsNullOrEmpty(message.RejectionInfo))
+                            {
+                                message.RejectionInfo = "Unable to send request - no rejection info available";
+                            }
+                            rejection = new OrleansMessageRejectionException(message.RejectionInfo);
+                        }
+                        break;
+                }
+                response = Response.ExceptionResponse(rejection);
+            }
+
+            if (!response.ExceptionFlag)
+            {
+                context.TrySetResult(response.Data);
+            }
+            else
+            {
+                context.TrySetException(response.Exception);
+            }
+        }
+
+        private static String GetDebugContext(string interfaceName, string methodName, object[] arguments)
+        {
+            // String concatenation is approx 35% faster than string.Format here
+            //debugContext = String.Format("{0}:{1}()", this.InterfaceName, GetMethodName(this.InterfaceId, methodId));
+            var debugContext = new StringBuilder();
+            debugContext.Append(interfaceName);
+            debugContext.Append(":");
+            debugContext.Append(methodName);
+            if (USE_DEBUG_CONTEXT_PARAMS && arguments != null && arguments.Length > 0)
+            {
+                debugContext.Append("(");
+                debugContext.Append(Utils.EnumerableToString(arguments));
+                debugContext.Append(")");
+            }
+            else
+            {
+                debugContext.Append("()");
+            }
+            return debugContext.ToString();
+        }
+
+        private static void CheckForGrainArguments(object[] arguments)
+        {
+            foreach (var argument in arguments)
+                if (argument is Grain)
+                    throw new ArgumentException(String.Format("Cannot pass a grain object {0} as an argument to a method. Pass this.AsReference<GrainInterface>() instead.", argument.GetType().FullName));
+        }
+
+        /// <summary>
+        /// Sets target grain to the found instances of type GrainCancellationToken
+        /// </summary>
+        /// <param name="arguments"> Grain method arguments list</param>
+        /// <param name="target"> Target grain reference</param>
+        private static void SetGrainCancellationTokensTarget(object[] arguments, GrainReference target)
+        {
+            if (arguments == null) return;
+            foreach (var argument in arguments)
+            {
+                (argument as GrainCancellationToken)?.AddGrainReference(target);
+            }
+        }
+
+        private bool IsUnordered(GrainReference reference)
+        {
+            return this.RuntimeClient.GrainTypeResolver?.IsUnordered(reference.GrainId.TypeCode) == true;
```

Comment:
Yes, was planning to do a perf run. Although you think it could impact perf at all? It's O(1)